### DISCUSSION TITLE
How to transfet All Data (%value%) with Rules in Tasmota (Sonoff ZB Bridge)

### DISCUSSION THREAD

**Discussion by alain16** [arendst/Tasmota#20832] [Q&A] [answered]
Hello,

I'm trying to use Rules with Tasmota on a Sonoff Zigbee Bridge to send a message to Domoticz with all Datas.
Then, I will use a script in Domoticz to decode the Data.
So I subscribed on the event ZbReceived.

What the sonoff ZBBridge console show : 
13:30:11.056 MQT: tele/SoBridge-Zigbee/SENSOR = {"ZbReceived":{"0x9B8B":{"Device":"0x9B8B","Name":"Room1","Humidity":68.8,"Endpoint":1,"LinkQuality":76}}}

Then I created a Rule to publish on domoticz with MQTT :
Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" : "ZbReceived"} endon

This send a "customEvent" with Mqtt : domoticz/in = {"command" : "customevent", "event" : "ZigBeeAction" , "data" : "ZbReceived"}

I well receive the event in Domoticz but I cannot transfer as a string the Data : 
{"0x9B8B":{"Device":"0x9B8B","Name":"Room1","Humidity":68.8,"Endpoint":1,"LinkQuality":76}}

My issue is how to send the Data as bloc in my customEvent ?

I don't want to do the script in Tasmota Rules because I have many different devices and this will need many Rules (difficult to maintain).

I have tried several Rule alternative as :
 Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" : "%value%"} endon
 Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" : %topic%} endon
 Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" : %prefix%/%topic%/} endon
 Rule1 
 on ZbReceived do Var1 %value% endon
 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" : "%Var1%"} endon

Helps will be very appreciated :)

Thanks, Alain

**Comment by sfromis** [comment]
With an ESP32-based bridge supporting the Berry language, you do have features for processing the whole JSON en bloc in Tasmota, but currently not with rules. This year, barbudor was trying to allow a `%json%` rule variable, but that did not work well, and IDK if the idea was dropped.

With any device able to process MQTT subscriptions, you could also do your bridging, including another Tasmota device with ESP32 and Berry.

**Reply by alain16** [reply]
Thanks sfromis !
I spend more than 6 hours in search :(
Unfortunately, I have a Sonoff Zigbee Bridge (ESP8266 + EFR32 with EZSP).

Trying another way :)
I saw that I can access to the Id of the device with %ZBDEVICE%

But my rule : 
Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  %ZBDEVICE%} endon
send : domoticz/in = {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  0xF553}
where 0xF553 is the DeviceID that triggered the rule.

How can I add the quote "0xF553" ?
How can I build a tab for "data" like this : {"0xF553", "data2", "data3"} ?

I aslo saw that it is possible to do a rule with a trigger : 
ZBReceived#?#Power=0 will trigger on {"ZBReceived":{"0x4773":{"Power":0}}} and on {"ZBReceived":{"aqara_switch":{"Power":0}}} both.

Is there something equivalent to %ZBDEVICE% to build my own tab of data like that:
domoticz/in = {"command" : "customevent", "event" : "ZigBeeAction" , "data" : MyOwnTab}
where MyOnwTab = { "%ZBDEVICE%", "ZBReceived#?#Power", "ZBReceived#?#Temperature", ... } 

Thanks for every helps 👍

**Reply by sfromis** [reply]
Not very clear what you're in doubt about, like what you mean by "something equivalent" or "own tab", maybe that it simpler than you think....

As you know about those rule variables `%zbdevice%` etc, they are available anywhere you'd like in a rule based on a Zigbee event, no matter what you do. And when you do a `Publish`, Tasmota knows nothing about what you put in the payload, and does not care about where you place quotes or curly brackets. You just construct your string in full, and then you can use `%` variables as placeholders where you have something varying.

What you **cannot** do is to use the trigger expressions with `#` anywhere but in the trigger part of a rule, between `on` and `do`. Anything you want to use, outside of the predefined `%zbxxxx%` variables, must be picked up in separate rules.

Yeah, that may be a pain in the .... which is why I pointed at processing the data outside of rules, which is just not made for more involved stuff.

BTW. If you know something about programming, forget about that when it comes to rules. It is a very different animal.

**Reply by alain16** [reply]
Sorry, for my french way of writing English. I wrote Tab instead of Array :)
Yes, I did some programing long time ago.

Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  "%ZBDEVICE%"} endon
work well.

Now I want to send an Array. 
To test, I wrote: 
Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  "{ %ZBDEVICE%, %topic%, %time% }"} endon
or
Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  "[ %ZBDEVICE%, %topic%, %time% ]"}"

In Domoticz the Event:
return {
    on = {customEvents = {"ZigBeeAction"}}, execute = function(domoticz, Message)
    print ('--------------------------    ZigBee Sensor Activated ');
    print ('--------------------------    Message[1]: '..Message.rawData[1]);
    print ('--------------------------    Message[2]: '..Message.rawData[2]);
    print ('--------------------------    Message[3]: '..Message.rawData[3]);
end
}
I do have the trace : Message.rawData[1], Message.rawData[2], Message.rawData[3]
Maybe I do not understand how Array are managed with MQTT.
Thanks!

**Reply by sfromis** [reply]
If you want something which Domoticz would treat as an array, you should probably just craft your [JSON](https://www.json.org/json-en.html) that way, which normally simply is a list of comma-delimited values enclosed in [] - remember `"` around values which are not plain numbers, but not around the whole list

This has nothing at all to do with MQTT, but is about the JSON data format commonly used on Internet.

**Comment by alain16** [comment]
As I said, programing was long time ago in C++, Fortan or x86 assembly :D

Based on your comment I'm testing:
Rule1 on ZbReceived do publish domoticz/in {"command" : "customevent", "event" : "ZigBeeAction" , "data" :  ["%ZBDEVICE%" , "%topic%" , "%time%" , "%var1%" , "%var2%" , "%var3%"] } endon

As I work in remote and I need to wait for an sensor activation, don't hesitate to check if my Rule syntax is good :)

Thank you

**Comment by alain16** [comment, marked as answer]
I found the solution 👍 

Rule in Tasmota Console on Sonoff Zigbee Bridge :

`Rule1
 ON ZbReceived#?#Temperature DO publish domoticz/in {"command":"customevent","event":"ZigBeeAction","data":"{\"idx\":1000,\"SensorId\":\"%ZBDEVICE%\",\"SensorType\":\"Temperature\",\"SensorValue\":\"%value%\"}" } ENDON
 ON ZbReceived#?#Humidity DO publish domoticz/in {"command":"customevent","event":"ZigBeeAction","data":"{\"idx\":1001,\"SensorId\":\"%ZBDEVICE%\",\"SensorType\":\"Humidity\",\"SensorValue\":\"%value%\"}" } ENDON
 ON ZbReceived#?#Contact DO publish domoticz/in {"command":"customevent","event":"ZigBeeAction","data":"{\"idx\":1002,\"SensorId\":\"%ZBDEVICE%\",\"SensorType\":\"Contact\",\"SensorValue\":\"%value%\"}" } ENDON
 ON ZbReceived#?#Occupancy DO publish domoticz/in {"command":"customevent","event":"ZigBeeAction","data":"{\"idx\":1003,\"SensorId\":\"%ZBDEVICE%\",\"SensorType\":\"Occupancy\",\"SensorValue\":\"%value%\"}" } ENDON
 ON ZbReceived#?#BatteryPercentage DO publish domoticz/in {"command":"customevent","event":"ZigBeeAction","data":"{\"idx\":1004,\"SensorId\":\"%ZBDEVICE%\",\"SensorType\":\"BatteryPercentage\",\"SensorValue\":\"%value%\"}" } ENDON 
Rule1 1
`

Then on Domoticz, I have the event:
`return {

	-- Manage a ZigbeeBridge Sensor

    on = {customEvents = {"ZigBeeAction"}}, execute = function(domoticz, triggeredItem)

    print ('--------------------------  xxx  A ZigBee sensor is activated ');

    if triggeredItem == nil then 						print('  x---  triggeredItem == nil');
	elseif triggeredItem.isCustomEvent == nil then 		print('  x---  CustomEnvent == nil');
	elseif triggeredItem.json == nil then 				print('  x---  triggeredItem.json == nil');
	elseif triggeredItem.json.IdCapteur == nil then 	print('  x---  triggeredItem.json.SensorID == nil');
    else
        print('Sensor Zigbee idx : '..triggeredItem.json.idx.. ', SensorId = '..triggeredItem.json.SensorId.. ', SensorType = '..triggeredItem.json.SensorType.. ', SensorValue = '..triggeredItem.json.SensorValue) ;

        -- Write your traitement code here

    end
    print ('--------------------------  End ZigBee sensor activated');
end
}
`

**Reply by sfromis** [reply]
So you decided to take the "pain in the ...." with a bunch of separate rules, as I mentioned. Of course, it is not what you asked for with "All Data", but only selected fields. But great if it works for you.

**Comment by Mibeus** [comment]
Hello,
I make plugin for domoticz for zigbee2tasmota
https://github.com/Mibeus/domoticz_zigbee2tasmota

you can check it
